planner/cascades/transformation_rules.go

```diff
@@ -634,3 +635,50 @@ func (r *EliminateProjection) OnTransform(old *memo.ExprIter) (newExprs []*memo.
 	}
 	return finalGroupExprs, true, false, nil
 }
+
+// MergeAdjacentProjection merge the adjacent projection.
+type MergeAdjacentProjection struct {
+	baseRule
+}
+
+// NewRuleMergeAdjacentProjection creates a new Transformation MergeAdjacentProjection.
+// The pattern of this rule is `Projection -> Projection`.
+func NewRuleMergeAdjacentProjection() Transformation {
+	rule := &MergeAdjacentProjection{}
+	rule.pattern = memo.BuildPattern(
+		memo.OperandProjection,
+		memo.EngineTiDBOnly,
+		memo.NewPattern(memo.OperandProjection, memo.EngineTiDBOnly),
+	)
+	return rule
+}
+
+// OnTransform implements Transformation interface.
+// It will transform `proj->proj->x` to `proj->x`
+// or just keep the adjacent projections unchanged.
+func (r *MergeAdjacentProjection) OnTransform(old *memo.ExprIter) (newExprs []*memo.GroupExpr, eraseOld bool, eraseAll bool, err error) {
+	proj := old.GetExpr().ExprNode.(*plannercore.LogicalProjection)
+	childGroup := old.Children[0].Group
+	child := old.Children[0].GetExpr().ExprNode.(*plannercore.LogicalProjection)
+	if plannercore.ExprsHasSideEffects(child.Exprs) {
+		return nil, false, false, nil
+	}
+
+	replace := make(map[string]*expression.Column)
+	for i, col := range childGroup.Prop.Schema.Columns {
+		if colOrigin, ok := child.Exprs[i].(*expression.Column); ok {
+			replace[string(col.HashCode(nil))] = colOrigin
+		}
+	}
+
+	newProj := plannercore.LogicalProjection{Exprs: proj.Exprs}.Init(proj.SCtx(), proj.SelectBlockOffset())
+	newProj.SetSchema(old.GetExpr().Group.Prop.Schema)
+	for i, expr := range newProj.Exprs {
+		plannercore.ResolveExprAndReplace(expr, replace)
```

Comment:
But the method `ResolveExprAndReplace` will change the `expr`? Then it will affect the origin `proj`. Please check this @winoros .

Reply:
Currently, the `ResolveExprAndReplace ` changed two things:
- the `RetType` is only changed when eliminating the `Aggregation` which caused bug https://github.com/pingcap/tidb/issues/13027 of course.
- the `InOperand` is only used for `NOT IN` subquery which can be removed from the `Column` struct.
This method should be removed soon once #13027 is fixed and `InOperand` is moved out.

I'm going to move the `InOperand` out.

Reply:
So we don't need to `Clone` an `expr` here?

Reply:
Yeah, i think it's safe enough.

Reply:
@pingyu There's a in-place modify in `replaceColumnOfExpr`. You can `Clone` the expr first to avoid modify the old proj's elements. We'll fix it later.
Sorry for your inconvenience.

Reply:
Never mind.
Updated, PTAL, thanks~ @winoros